synapse/cryotank.py

```diff
@@ -571,3 +588,610 @@ def init(self, name, conf=None, timeout=None):
+                        if field is None:
+                            continue
+                        # TODO : what to do with subprops?
+                        break
+                    else:
+                        logger.debug('Datapaths %s yield nothing for offset %d', idx.datapath, offset)
+                        continue
+                    normval, _ = s_datamodel.getTypeNorm(idx.syntype, field)
+                except (s_exc.NoSuchType, s_exc.BadTypeValu):
+                    logger.debug('Norm fail')
+                    self._meta.progresses[iid]['nnormfail'] += 1
+                    continue
+                self._meta.progresses[iid]['ngood'] += 1
+                yield offset, iid, normval
+
+    def _writeIndices(self, rows):
+        '''
+        Persists actual indexing to disk.
+
+        Args:
+            rows(Iterable[Tuple[int, int, Union[str, int]]]):  generators of tuples of offset, index ID,  normalized
+            property value
+
+        Returns:
+            int:  the next cryotank offset that should be indexed
+        '''
+        count = -1
+        with self._dbenv.begin(db=self._idxtbl, buffers=True, write=True) as txn:
+            logger.debug('_dbenv.begin(a, buffers=True, write=True')
+            for count, (offset, iid, normval) in enumerate(rows):
+
+                offset_enc = _Int64be.pack(offset)
+                iid_enc = _iid_en(iid)
+                valkey_enc = s_lmdb.encodeValAsKey(normval)
+
+                logger.debug('txn.put(%r, %r)', iid_enc + valkey_enc, offset_enc)
+                txn.put(iid_enc + valkey_enc, offset_enc)
+                txn.put(offset_enc + iid_enc, s_msgpack.en(normval), db=self._normtbl)
+
+            self._meta.persist(progressonly=True, txn=txn)
+            logger.debug('txn end')
+        return count + 1
+
+    def _workerloop(self):
+        stillworktodo = True
+
+        while True:
+            # Run the outstanding commands
+            recalc = False
+            while True:
+                try:
+                    job = self._workq.get(timeout=0 if stillworktodo else None)
+                    stillworktodo = True
+                    retn, callback, args, kwargs = job
+                    try:
+                        if retn is not None:
+                            retn.retn(callback(*args, **kwargs))
+                            recalc = True
+                    except Exception as e:
+                        if retn is None:
+                            raise
+                        else:
+                            # Not using errx because I want the exception object itself
+                            retn.retn(e)
+                except s_exc.IsFini:
+                    return
+                except s_exc.TimeOut:
+                    break
+            if recalc:
+                self._next_offset = self._meta.lowestProgress()
+
+            # loop_start_t = time.time()
+            import itertools
```

Comment:
should we just import this at the top?

Reply:
oh good catch. That tee stuff shouldn't be in there at all.